app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+    ssa_profile = iam.instance_profile(profile_name)
+    ssa_profile = iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    find_or_create_role(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.zero?
+
+    ssa_profile
+  end
+
+  # possible RHEL image name: values: [ "RHEL-7.3_HVM_GA*" ]
+  def get_agent_image_id(img = agent_ami)
+    imgs = ec2.client.describe_images(
+      filters: [
+        {
+          name: "name",
+          values: [img]
+        }
+      ]
+    ).images
+
+    imgs[0].image_id
+  end
+
+  def create_security_group(group_name = MIQ_SSA)
+    sgs = ec2.client.describe_security_groups(
+      filters: [
+        {
+          name: "group-name",
+          values: [group_name]
+        }
+      ]
+    ).security_groups
+    return sgs[0].group_id unless sgs.empty?
+
+    # create security group if not exist
+    security_group = ec2.create_security_group({
+      group_name: group_name,
+      description: 'Security group for MIQ SSA Agent',
+      vpc_id: ec2.client.describe_vpcs.vpcs[0].vpc_id
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 22,
+        to_port: 22,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 80,
+        to_port: 80,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 443,
+        to_port: 443,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.group_id
+  end
+
+  # Get Key Pair for SSH. Create a new one if not exists.
+  def get_key_pair(pair_name = MIQ_SSA)
+    kps = Authentication.where(name: pair_name)
+    kps.each do |kp|
+      return kp if kp.resource_id == @ems.id
+    end
+
+    ManageIQ::Providers::CloudManager::AuthKeyPair.create_key_pair(@ems.id,
+      { :key_name => pair_name })
+  end
+
+  def create_pem_file(pair_name = MIQ_SSA)
+    kp = get_key_pair(pair_name)
+    pem_file_name = pair_name + ".pem"
+    File.open(pem_file_name, 'w') { |f| f.write(kp.auth_key) }
+    File.chmod(0400, pem_file_name)
+    pem_file_name
+  end
+
+  def create_setting_yaml(yml = "tools/amazon_agent_settings/default_ssa_config.yml")
+    defaults = Hash(agent_manager_settings)
+    defaults[:region] = region
+    defaults[:request_queue] = request_queue
+    defaults[:reply_queue] = reply_queue
+    defaults[:ssa_bucket] = ssa_bucket
+    File.open(yml, "w") { |f| f.write(defaults.to_yaml) }
+  end
+
+  def create_userdata
+    File.chmod(0755, userdata_script)
+    stdout, stderr, status = Open3.capture3(userdata_script, "#{ruby_version}", "#{log_level}")
```

Comment:
I'm confused, I thought this library runs on the worker side...why would it shell out the userdata_script?